### PR TITLE
feat(form ): simple/advanced form implementation

### PR DESCRIPTION
**Pull request by mJarsater**
* Implemented a simple form, using a <form>.
* Implemented a advanced form, using Reactive form

### CODE REVIEW

**Comment by mJarsater** [comment]
> Great work! LGTM.

@gokberknur I pushed a commit just now when I use a directive for the checkbox instead of handling everything in the (tdsChange) listener. My thinking is that this approach is much more scalable and generic, people could possible just copy/paste this directive and they would be up and running.

Would you say this approach is better, weirder, the same? What does you Angular-eyes say? 

Any feedback is super valuable!

**Comment by gokberknur** [comment]
> > Great work! LGTM.
> 
> @gokberknur I pushed a commit just now when I use a directive for the checkbox instead of handling everything in the (tdsChange) listener. My thinking is that this approach is much more scalable and generic, people could possible just copy/paste this directive and they would be up and running.
> 
> Would you say this approach is better, weirder, the same? What does you Angular-eyes say?
> 
> Any feedback is super valuable!

@mJarsater @theJohnnyMe 
Great thought! I see where you are going and I like it the way you are making things. However,

In this example I think just for handling (tdsChange) it might be overkill. 

directives are useful for patching behaviour of the elements, especially if we need some sort of ccustom/complex logic for specific elements that we have.

In this example, this directive applies for all checkboxes in the application and handling checkbox values provided via FormGroup. While this makes sense for this use case, it is tricky to know what user will do with their checkbox element. 
I would suggest give different name to directive to explicitly separate this use case. Because right now this logic will apply to all tds-checkbox elements.

In some cases for example, user may want to emit some events in their components (to parent component), call another function, or bind some reactivity to that checkbox to create some side effects. 

I would suggest to have directives on special use cases. e.g: dropdown, because it is obvious that we need to read/write value of the dropdown from our angular component which makes necessary to have directive for that (custom-elements in general)


Another special use case might be handling [state] (error messages) in forms. This can save the user re-write, and I think it is generally applicabile because all it needs to do is show the error state or other state of the form:

`[state]="showErrorLogic ? 'error' : 'success' ` 


To sum up, it would be nice to demonstrate how to use directive on tegel elements but would be even nicer to specify in which case you should use/or not, especially for patching native events.

Let me know your thoughts :)

**Comment by mJarsater** [comment]
> > > Great work! LGTM.
> > 
> > 
> > @gokberknur I pushed a commit just now when I use a directive for the checkbox instead of handling everything in the (tdsChange) listener. My thinking is that this approach is much more scalable and generic, people could possible just copy/paste this directive and they would be up and running.
> > Would you say this approach is better, weirder, the same? What does you Angular-eyes say?
> > Any feedback is super valuable!
> 
> @mJarsater @theJohnnyMe Great thought! I see where you are going and I like it the way you are making things. However,
> 
> In this example I think just for handling (tdsChange) it might be overkill.
> 
> directives are useful for patching behaviour of the elements, especially if we need some sort of ccustom/complex logic for specific elements that we have.
> 
> In this example, this directive applies for all checkboxes in the application and handling checkbox values provided via FormGroup. While this makes sense for this use case, it is tricky to know what user will do with their checkbox element. I would suggest give different name to directive to explicitly separate this use case. Because right now this logic will apply to all tds-checkbox elements.
> 
> In some cases for example, user may want to emit some events in their components (to parent component), call another function, or bind some reactivity to that checkbox to create some side effects.
> 
> I would suggest to have directives on special use cases. e.g: dropdown, because it is obvious that we need to read/write value of the dropdown from our angular component which makes necessary to have directive for that (custom-elements in general)
> 
> Another special use case might be handling [state] (error messages) in forms. This can save the user re-write, and I think it is generally applicabile because all it needs to do is show the error state or other state of the form:
> 
> `[state]="showErrorLogic ? 'error' : 'success' `
> 
> To sum up, it would be nice to demonstrate how to use directive on tegel elements but would be even nicer to specify in which case you should use/or not, especially for patching native events.
> 
> Let me know your thoughts :)

Great! I will revert some of the changes for the checkbox and slider, however do you feel like a directive for the dropdown is necessary? As far as I can tell it would also just need to update the value on change?

**Comment by gokberknur** [comment]
> > > > Great work! LGTM.
> > > 
> > > 
> > > @gokberknur I pushed a commit just now when I use a directive for the checkbox instead of handling everything in the (tdsChange) listener. My thinking is that this approach is much more scalable and generic, people could possible just copy/paste this directive and they would be up and running.
> > > Would you say this approach is better, weirder, the same? What does you Angular-eyes say?
> > > Any feedback is super valuable!
> > 
> > 
> > @mJarsater @theJohnnyMe Great thought! I see where you are going and I like it the way you are making things. However,
> > In this example I think just for handling (tdsChange) it might be overkill.
> > directives are useful for patching behaviour of the elements, especially if we need some sort of ccustom/complex logic for specific elements that we have.
> > In this example, this directive applies for all checkboxes in the application and handling checkbox values provided via FormGroup. While this makes sense for this use case, it is tricky to know what user will do with their checkbox element. I would suggest give different name to directive to explicitly separate this use case. Because right now this logic will apply to all tds-checkbox elements.
> > In some cases for example, user may want to emit some events in their components (to parent component), call another function, or bind some reactivity to that checkbox to create some side effects.
> > I would suggest to have directives on special use cases. e.g: dropdown, because it is obvious that we need to read/write value of the dropdown from our angular component which makes necessary to have directive for that (custom-elements in general)
> > Another special use case might be handling [state] (error messages) in forms. This can save the user re-write, and I think it is generally applicabile because all it needs to do is show the error state or other state of the form:
> > `[state]="showErrorLogic ? 'error' : 'success' `
> > To sum up, it would be nice to demonstrate how to use directive on tegel elements but would be even nicer to specify in which case you should use/or not, especially for patching native events.
> > Let me know your thoughts :)
> 
> Great! I will revert some of the changes for the checkbox and slider, however do you feel like a directive for the dropdown is necessary? As far as I can tell it would also just need to update the value on change?

wait actually you have a point. There are 2 checkbox components and i was checking the wrong one!!! 

Thinking out loud now. 

Maybe it would be better to create a use-case? to show re-usability of directive. 

If we want to bind angular forms (reactive and template driven)with tegel elements, thats where we need to use NG_VALUE_ACCESSOR (in this example via directive) because it provides bridge between custom-elements and angular forms https://angular.io/api/forms/ControlValueAccessor 

thats sort of what ngdefaultcontrol does to <input> fields. 

but sometimes, we may need more complex use-cases (and re-usable ones)

maybe we can create seperate field in the demo page for angular form driven more complex usage of elements and simple usage of elements?

we need to rename the directive selectors tho. (if we name it like tds-dropdown, it means it will apply all dropdowns in the entire application, which is not we want)

example:

so if you want to modify dropdown behaviour go and check dropdown-handler directive, and do your magic there and attach it to dropdown-element. (and if you are going to use this magic for other parts of the application on your dropdowns, there you go)

if you need to just show selected value of dropdown without any logic, go ahead and use normal element.

 maybe this would help people to understand simple usage and other complex scenarios? 

what do you think?

**Comment by mJarsater** [comment]
> > > > > Great work! LGTM.
> > > > 
> > > > 
> > > > @gokberknur I pushed a commit just now when I use a directive for the checkbox instead of handling everything in the (tdsChange) listener. My thinking is that this approach is much more scalable and generic, people could possible just copy/paste this directive and they would be up and running.
> > > > Would you say this approach is better, weirder, the same? What does you Angular-eyes say?
> > > > Any feedback is super valuable!
> > > 
> > > 
> > > @mJarsater @theJohnnyMe Great thought! I see where you are going and I like it the way you are making things. However,
> > > In this example I think just for handling (tdsChange) it might be overkill.
> > > directives are useful for patching behaviour of the elements, especially if we need some sort of ccustom/complex logic for specific elements that we have.
> > > In this example, this directive applies for all checkboxes in the application and handling checkbox values provided via FormGroup. While this makes sense for this use case, it is tricky to know what user will do with their checkbox element. I would suggest give different name to directive to explicitly separate this use case. Because right now this logic will apply to all tds-checkbox elements.
> > > In some cases for example, user may want to emit some events in their components (to parent component), call another function, or bind some reactivity to that checkbox to create some side effects.
> > > I would suggest to have directives on special use cases. e.g: dropdown, because it is obvious that we need to read/write value of the dropdown from our angular component which makes necessary to have directive for that (custom-elements in general)
> > > Another special use case might be handling [state] (error messages) in forms. This can save the user re-write, and I think it is generally applicabile because all it needs to do is show the error state or other state of the form:
> > > `[state]="showErrorLogic ? 'error' : 'success' `
> > > To sum up, it would be nice to demonstrate how to use directive on tegel elements but would be even nicer to specify in which case you should use/or not, especially for patching native events.
> > > Let me know your thoughts :)
> > 
> > 
> > Great! I will revert some of the changes for the checkbox and slider, however do you feel like a directive for the dropdown is necessary? As far as I can tell it would also just need to update the value on change?
> 
> wait actually you have a point. There are 2 checkbox components and i was checking the wrong one!!!
> 
> Thinking out loud now.
> 
> Maybe it would be better to create a use-case? to show re-usability of directive.
> 
> If we want to bind angular forms (reactive and template driven)with tegel elements, thats where we need to use NG_VALUE_ACCESSOR (in this example via directive) because it provides bridge between custom-elements and angular forms https://angular.io/api/forms/ControlValueAccessor
> 
> thats sort of what ngdefaultcontrol does to fields.
> 
> but sometimes, we may need more complex use-cases (and re-usable ones)
> 
> maybe we can create seperate field in the demo page for angular form driven more complex usage of elements and simple usage of elements?
> 
> we need to rename the directive selectors tho. (if we name it like tds-dropdown, it means it will apply all dropdowns in the entire application, which is not we want)
> 
> example:
> 
> so if you want to modify dropdown behaviour go and check dropdown-handler directive, and do your magic there and attach it to dropdown-element. (and if you are going to use this magic for other parts of the application on your dropdowns, there you go)
> 
> if you need to just show selected value of dropdown without any logic, go ahead and use normal element.
> 
> maybe this would help people to understand simple usage and other complex scenarios?
> 
> what do you think?

Super input!

I agree that we can create a couple of different form with different level of complexity. I will do so on out form page and try to mark them as "simple form", "advanced form". 

I will also try and differentiate the directives and apply them only to the components that should "use" them.

**Comment by gokberknur** [comment]
To clarify: we (me and tommy) checked the implementaiton, we have some suggestions currently WIP.

**Comment by mJarsater** [comment]
> To clarify: we (me and tommy) checked the implementaiton, we have some suggestions currently WIP.

Sweet, do you wanna add a commit to this PR whenever you feel ready?

File: src/app/pages/form/dropdown/dropdown.component.html [hunk 1]
```diff
@@ -37,18 +34,15 @@
       open-direction="down"
 
     >
-      <ng-container *ngFor="let town of nowegianTowns; let idx = index">
-        <tds-dropdown-option [value]="town.value">
-          {{ town.label }}
-        </tds-dropdown-option>
-      </ng-container>
+      <tds-dropdown-option *ngFor="let town of nowegianTowns; let idx = index" value="{{town.value}}">
```
Comment (mJarsater) [inline src/app/pages/form/dropdown/dropdown.component.html:37]: Nice with the value in the *ngFor 🍻

Comment (gokberknur) [reply src/app/pages/form/dropdown/dropdown.component.html:37]: Yes, ng-container is not needed here. :)

**Comment by gokberknur** [comment]
ups i creaated some conflicts :) 

hmmm.  maybe we can solve together after you finish all other on-going work. 

there are some changes in routing module. (lazy-loading components). 

when authoring standalone components, if we want them to be lazy-loaded, "default" needs to be used before exporting classes.

and thats all.

**Comment by mJarsater** [comment]
> ups i creaated some conflicts :)
> 
> hmmm. maybe we can solve together after you finish all other on-going work.
> 
> there are some changes in routing module. (lazy-loading components).
> 
> when authoring standalone components, if we want them to be lazy-loaded, "default" needs to be used before exporting classes.
> 
> and thats all.

Merge conflicts are solved now. Saw that you added a tab for template driven form, do you want to do this implementation also?
